File: 백준/Gold/10282. 해킹/해킹.py
import sys
import heapq

INF = int(1e9)

def dijkstra(start, n, graph):
    dist = [INF] * (n + 1)
    dist[start] = 0
    pq = [(0, start)]  
    
    while pq:
        time, node = heapq.heappop(pq)
        
        if time > dist[node]:
            continue
        
        for next_node, next_time in graph[node]:
            if dist[next_node] > time + next_time:
                dist[next_node] = time + next_time
                heapq.heappush(pq, (dist[next_node], next_node))

    infected = sum(1 for d in dist if d != INF)
    max_time = max((d for d in dist if d != INF), default=0)
    
    print(infected, max_time)

def main():
    input = sys.stdin.read
    data = input().splitlines()
    
    t = int(data[0])
    index = 1
    
    for _ in range(t):
        n, d, c = map(int, data[index].split())
        index += 1
        graph = [[] for _ in range(n + 1)]
        for _ in range(d):
            a, b, s = map(int, data[index].split())
            graph[b].append((a, s))
            index += 1
        
        dijkstra(c, n, graph)

if __name__ == "__main__":
    main()

File: 백준/Gold/10282. 해킹/해킹.cc
#include <bits/stdc++.h>
using namespace std;
using pii = pair<int, int>;

const int INF = 1e9;
vector<pii> g[10001];

void dijkstra(int start, int n) {
    priority_queue<pii, vector<pii>, greater<pii>> pq;
    vector<int> dist(n + 1, INF);
    dist[start] = 0;
    pq.push({0, start});

    while (!pq.empty()) {
        auto [time, node] = pq.top();
        pq.pop();

        if (time > dist[node]) continue;

        for (auto &[nextNode, nextTime] : g[node]) {
            if (dist[nextNode] > time + nextTime) {
                dist[nextNode] = time + nextTime;
                pq.push({dist[nextNode], nextNode});
            }
        }
    }

    int infected = 0, maxTime = 0;
    for (int i = 1; i <= n; i++) {
        if (dist[i] != INF) {
            infected++;
            maxTime = max(maxTime, dist[i]);
        }
    }
    cout << infected << " " << maxTime << "\n";
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, d, c;
        cin >> n >> d >> c;
        for (int i = 0; i <= n; i++) g[i].clear();

        for (int i = 0; i < d; i++) {
            int a, b, s;
            cin >> a >> b >> s;
            g[b].push_back({a, s});
        }

        dijkstra(c, n);
    }
    return 0;
}